### ISSUE TITLE
Low validation acc with BatchNorm, but CuDNNBatchNorm is fine

### ISSUE THREAD

**Issue by neodooth** [dmlc/mxnet#2553]
Hi, I built mxnet with CUDA 7.0 and CuDNN v4, and trained inception-bn for classification. However, if I use BatchNorm, the training accuracy is fine(17%), but the validation accuracy is unreasonably low (3%), and become even lower after some more epochs (like 0.003%). If I replace BatchNorm with CuDNNBatchNorm, everything works fine.

Should I use CuDNNBatchNorm if mxnet is built with CuDNN? And what's the difference between BN and CuDNNBN? Also I noticed there is CuDNNConvolution, but normal Convolution works good for me. It's quite confusing.

And, if I use an older version of mxnet (commit f91b2e51725aef700d619d8a605bbd6f437910bb, Mar 7), normal BN is OK.

Thanks

**Comment by sxjscience** [comment]
@piiswrong Seems that the BatchNorm will fall into the CuDNN version only if using CuDNN V5? https://github.com/dmlc/mxnet/blob/master/src/operator/batch_norm.cu#L15.

**Comment by piiswrong** [comment]
@antinucleon sounds like our bn is wrong?

@neodooth This is probably due to the madness with cudnn versioning. They did some non compatible changes.

If you use cudnn v5 and newest mxnet, BatchNorm should use cudnn by default.

**Comment by antinucleon** [comment]
@piiswrong I checked with @winstywang  before but we didn't find out anything. I will check it once more after finish current work.

**Comment by ommiissyu** [comment]
I also find the same problem. If I use the cudnn v4, the gpu and cpu's batchnorm results are the same, but if I use the cudnn v5, the gpu and cpu's batchnorm results are totally different.

**Comment by piiswrong** [comment]
Looks like our batchnorm implementation is wrong. Should fix it

**Comment by piiswrong** [comment]
seems like fix_gamma problem. Should be fixed by https://github.com/dmlc/mxnet/pull/2563

**Comment by neodooth** [comment]
Yes, #2563 fixed this. Thanks!
